xoto3/dynamodb/exceptions.py

```diff
@@ -1,27 +1,75 @@
 """Exceptions for our Dynamo usage"""
+from typing import Optional, TypeVar, Type, Dict, Tuple
 import botocore.exceptions
 
+from .types import ItemKey
+
 
 class DynamoDbException(Exception):
     """Wrapping error responses from Dynamo DB"""
 
     pass
 
 
-class AlreadyExistsException(DynamoDbException):
+class DynamoDbItemException(DynamoDbException):
+    def __init__(self, msg: str, *, key: Optional[ItemKey] = None, table_name: str = "", **kwargs):
+        self.__dict__.update(kwargs)
+        self.key = key
+        self.table_name = table_name
+        super().__init__(msg)
+
+
+class AlreadyExistsException(DynamoDbItemException):
+    """Deprecated - prefer ItemAlreadyExistsException"""
+
+    pass
+
+
+class ItemAlreadyExistsException(AlreadyExistsException):
+    """Backwards-compatible, more consistent name"""
+
     pass
 
 
-class ItemNotFoundException(DynamoDbException):
+class ItemNotFoundException(DynamoDbItemException):
     """Being more specific that an item was not found"""
 
+    pass
+
+
+X = TypeVar("X", bound=DynamoDbItemException)
+
+
+_GENERATED_ITEM_EXCEPTION_TYPES: Dict[Tuple[str, str], type] = {
```

Comment:
At first read, this isn't immediately clear to be a cache of previously-assigned key-exception mappings. Perhaps that's because I'm predisposed to seeing all-caps as defining a constant. Setting this up as a mutable array threw me a bit...

Is the "ItemNotFoundException" included because it's a commonly used case and this is a bit of optimization?

Reply:
i like to name globals with all-caps names. Globals are scary and i like it when they look scary. :)

i mostly added it to actually have something that would help mypy 'verify' that my type annotation was sane.